### PR TITLE
layer-shell: add support for popups

### PR DESCRIPTION
**Pull request by ddevault**
Supports nested popups as well.

### CODE REVIEW

File: sway/desktop/layer_shell.c [hunk 1]
```diff
@@ -350,6 +350,114 @@ static void handle_unmap(struct wl_listener *listener, void *data) {
...
+	return popup->parent_layer;
+}
+
+static void popup_handle_map(struct wl_listener *listener, void *data) {
+	struct sway_layer_popup *popup = wl_container_of(listener, popup, map);
+	struct sway_layer_surface *layer = popup_get_layer(popup);
+	struct wlr_output *wlr_output = layer->layer_surface->output;
+	struct sway_output *output = wlr_output->data;
+	// TODO: Proper damage tracking
+	output_damage_whole(output);
```
Comment (emersion) [inline sway/desktop/layer_shell.c]: Err. We could at least only damage the popup.

File: sway/desktop/layer_shell.c [hunk 2]
```diff
@@ -350,6 +350,114 @@ static void handle_unmap(struct wl_listener *listener, void *data) {
...
+}
+
+static void popup_handle_commit(struct wl_listener *listener, void *data) {
+	struct sway_layer_popup *popup = wl_container_of(listener, popup, commit);
+	struct sway_layer_surface *layer = popup_get_layer(popup);
+	struct wlr_output *wlr_output = layer->layer_surface->output;
+	struct sway_output *output = wlr_output->data;
+	// TODO: Proper damage tracking
+	output_damage_whole(output);
```
Comment (emersion) [inline sway/desktop/layer_shell.c]: And here take damage from the parent surface, without damaging the whole thing.

**Comment by ddevault** [comment]
Investigating an issue with hidpi

**Comment by ddevault** [comment]
This should be ready to go.

File: sway/desktop/layer_shell.c [hunk 3]
```diff
@@ -350,6 +349,131 @@ static void handle_unmap(struct wl_listener *listener, void *data) {
...
+}
+
+static void popup_handle_unmap(struct wl_listener *listener, void *data) {
+	struct sway_layer_popup *popup = wl_container_of(listener, popup, unmap);
+	popup_damage(popup);
+}
+
+static void popup_handle_commit(struct wl_listener *listener, void *data) {
+	struct sway_layer_popup *popup = wl_container_of(listener, popup, commit);
+	popup_damage(popup);
```
Comment (emersion) [inline sway/desktop/layer_shell.c]: This damages the whole popup. It should only take damage from the surface:

```c
output_damage_surface(output, ox, oy, surface, false);
```

**Comment by ddevault** [comment]
Bump

**Comment by luispabon** [comment]
I'm testing this together with mate's wayland port of their panel. 

  * Add panel
  * Add application menu
  * Click on it 
  * Click away to close it

image

Similar stuff happens with context menues on the panel.

File: sway/desktop/output.c [hunk 4]
```diff
@@ -204,6 +204,36 @@ void output_layer_for_each_surface(struct sway_output *output,
 		output_surface_for_each_surface(output, wlr_layer_surface_v1->surface,
 			layer_surface->geo.x, layer_surface->geo.y, iterator,
 			user_data);
+
+		struct wlr_xdg_popup *state;
+		wl_list_for_each(state, &wlr_layer_surface_v1->popups, link) {
```
Comment (emersion) [inline sway/desktop/output.c]: We have `wlr_layer_surface_v1_for_each_surface` for this. But oh well, w/e

Comment (ddevault) [reply sway/desktop/output.c]: I was going to use this but there were Problems and an hour in I gave up.

File: sway/desktop/layer_shell.c [hunk 5]
```diff
@@ -350,6 +349,131 @@ static void handle_unmap(struct wl_listener *listener, void *data) {
...
+}
+
+static void popup_handle_destroy(struct wl_listener *listener, void *data) {
+	struct sway_layer_popup *popup =
+		wl_container_of(listener, popup, destroy);
+
+	wl_list_remove(&popup->map.link);
+	wl_list_remove(&popup->unmap.link);
+	wl_list_remove(&popup->destroy.link);
+	wl_list_remove(&popup->commit.link);
```
Comment (emersion) [inline sway/desktop/layer_shell.c]: Missing `new_popup`

File: sway/desktop/layer_shell.c [hunk 6]
```diff
@@ -350,6 +349,131 @@ static void handle_unmap(struct wl_listener *listener, void *data) {
...
+	struct sway_output *output = wlr_output->data;
+	output_damage_surface(output, ox, oy, surface, false);
+}
+
+static void popup_handle_map(struct wl_listener *listener, void *data) {
+	struct sway_layer_popup *popup = wl_container_of(listener, popup, map);
+	struct sway_layer_surface *layer = popup_get_layer(popup);
+	struct wlr_output *wlr_output = layer->layer_surface->output;
+	wlr_surface_send_enter(popup->wlr_popup->base->surface, wlr_output);
+	popup_damage(popup);
```
Comment (emersion) [inline sway/desktop/layer_shell.c]: This should damage the whole popup, but doesn't.

**Comment by Kommynct** [comment]
Is this going to make it into 1.2 or is it planned for later?

**Comment by emersion** [comment]
No, only bugfixes can make it. I was away so I couldn't get this reviewed in time for 1.2.

**Pull request merged**